### PR TITLE
kissfft/131.1.0: add more public defines

### PR DESCRIPTION
**Pull request by paulharris**
These were defined in kissfft's CMakeLists.txt, but they did not exist
in the recipe.

kissfft/131.1.0

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by paulharris** [comment]
Actually, surely I could use the pkg file that is generated?
Although, I don't see it built in the build folder...

**Comment by paulharris** [comment]
Question for the authors of #4935 
@SSE4 @SpaceIm 

the original recipe had KISSFFT_PKGCONFIG=False
which disabled the kissfft.pc file generation.

Shouldn't conan package_info() be using the generated .pc file to generate the correct component[].defines ?

Currently I have to eye-ball the program and manually add definitions to the recipe, which seems so very fragile.  We already have the required data exported in a .pc file, it should be used, right?

**Review by SpaceIm — APPROVED** [review]

File: recipes/kissfft/all/conanfile.py [hunk 1]
```diff
@@ -88,8 +88,23 @@ def package_info(self):
         # TODO: back to global scope in conan v2 once cmake_find_package_* generators removed
         self.cpp_info.components["libkissfft"].includedirs.append(os.path.join("include", "kissfft"))
         self.cpp_info.components["libkissfft"].libs = [lib_name]
+
+        # got to duplicate the logic from kissfft/CMakeLists.txt
+        if self.options.datatype == "float" or self.options.datatype == "double":
```
Comment (SpaceIm) [inline recipes/kissfft/all/conanfile.py:93]:
```suggestion
        if self.options.datatype in ["float", "double"]:
```

**Comment by paulharris** [comment]
It generates different .pc files depending on what features were enabled.
eg this is kissfft-double.pc (because I asked for datatype=double)

```
prefix=K:/env/conan2/deps/cci-fixes/conan-cci-kissfft/recipes/kissfft/all/package
libdir=${prefix}/lib
includedir=${prefix}/include

Name: kissfft
Description: A Fast Fourier Transform (FFT) library that tries to Keep it Simple, Stupid
Version: 131.1.0

Libs:  -L${libdir} -lkissfft-double
Cflags: -I${includedir}/kissfft  -Dkiss_fft_scalar=double
```

Note, this also will lead to a broader discussion about packaging libraries.  We have many libraries where we just specify the includes and libs, and I wonder how many need extra #defines that aren't aware of...

VTK is highly complicated too, I wonder how best to transfer the knowledge of flags + internal dependencies from VTK's cmake system into a recipe?
.pc files, or just extend their modules.json export (built during build) to include extra things the recipe needs to know?

**Comment by paulharris** [comment]
@SSE4 I had a look around, I don't understand how we can use pc / pkg.
I see libcurl exports a curl-config script, and
`curl-config --cflags` outputs the full absolute path...
I thought it wasn't supposed to use absolute paths?

And if kissfft switches to pkg, does that mean the consuming recipes have to be adjusted to import via pkg?

I had imagined we could dump the .pc files into the package folder, and the recipe's package_info() would use them to populate conan's defines, lib names, etc.
Else, how can the library's build system communicate what defines are required to its own recipe?

**Comment by SpaceIm** [comment]
> @SSE4 I had a look around, I don't understand how we can use pc / pkg. I see libcurl exports a curl-config script, and `curl-config --cflags` outputs the full absolute path... I thought it wasn't supposed to use absolute paths?
> 
> And if kissfft switches to pkg, does that mean the consuming recipes have to be adjusted to import via pkg?
> 
> I had imagined we could dump the .pc files into the package folder, and the recipe's package_info() would use them to populate conan's defines, lib names, etc. Else, how can the library's build system communicate what defines are required to its own recipe?

No, conan provides `pkg_config` generator for users who want to consume a library with pkg-config. Recipes just have to properly define consuming information in `package_info()`, it's not specific to pkg-config, I don't understand your reasoning.

**Comment by paulharris** [comment]
My goal is to use the information generated by the library's build to populate the recipe cpp_info.

eg, the #defines library consumers require is encoded into the .pc / pkgconfig that the library generates during build.
The #defines that the library specifies depends on many things within the library's own CMake system, and while often the logic is simple (eg CURL_STATIC=1) other times the logic is complex (kissfft:datatype = different types, translates into entirely different #defines).

And worse, if (eg) kissfft adds another type to datatype, it'll mean changes to the #defines that consumers will need to specify.  This may go unnoticed when the recipe incorporates the next version.

For kissfft, the #defines required are already encoded into its generated pkgconfig output.
Why can't conan capture those outputs for the cpp_info in package_info() ?

So this is not about generating a pkg for a consumer, it is about conan collecting build information.


Note that it would be better if kissfft simply generated a header with the appropriate #defines... Then we would've have to worry about capturing this stuff ourselves at all.
But if we can support it, then it would be easier to write recipes for libraries that we cannot adjust upstream.

**Comment by SpaceIm** [comment]
kissfft interface is simple, it's not worth the effort. If you want to see how it could work (with CMak config file, not pkgconfig), take a look at abseil recipe.

**Comment by paulharris** [comment]
>  If you want to see how you can populate package_info() from files generated at packaging time (with CMake config file as an input, not pkgconfig file), take a look at abseil recipe.

Thanks for this reference, I had this approach in my mind when considering the VTK recipe.  It also has defines and extra things added in.

This is really not a nice way of doing things.  It feels like there needs to be a neater way of transferring this kind of information from the build to the recipe.  Parsing the cmake file is fragile and complex.
VTK's modules.json is close.   .pc / pkgconfig is another option, but I feel not detailed enough for conan's needs.

I don't even see a good standard way that exists right now.  What can a cooperative upstream library do to export this kind of cpp_info information for conan to consume?
Perhaps a cmake 'export target info for conan' function that could be added to upstream library cmake?   Then it can generate a .json file, designed for conan, that can be processed in either the package() step, or kept in the package_folder and processed during the package_info() step.

**Comment by SSE4** [comment]
the proper way to go (at least for CMake projects) would be to use [CMake File API](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html) (more precisely, [codemodel v2](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html#codemodel-version-2)). see also https://github.com/conan-io/conan/pull/9005 - I've started an effort some time ago, but at this moment, CMake File API was limited and didn't provide all the information we needed. you can try to resume and continue from there.

**Review by prince-chrismc — APPROVED** [review]

**Comment by paulharris** [comment]
> the proper way to go (at least for CMake projects) would be to use [CMake File API](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html) (more precisely, [codemodel v2](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html#codemodel-version-2)). see also [conan-io/conan#9005](https://github.com/conan-io/conan/pull/9005) - I've started an effort some time ago, but at this moment, CMake File API was limited and didn't provide all the information we needed. you can try to resume and continue from there.

Ok thanks, I won't do it for kissfft, I'll keep it in mind for the future.

I'm ready to merge this one.

**Review by SSE4 — APPROVED** [review]

**Pull request merged**